compiler/noirc_frontend/src/hir_def/types.rs

```diff
@@ -3119,3 +3125,101 @@ impl PartialEq for Type {
         }
     }
 }
+
+/// Call a visitor function on the [Type] itself, and then each contained [Type],
+/// unless the visitor returned `false`, indicating that the type has been handled.
+pub(crate) fn visit_type<F>(typ: &Type, v: &mut F)
+where
+    F: FnMut(&Type) -> bool,
+{
+    if !v(typ) {
+        return;
+    }
+    match typ {
+        Type::FieldElement => {}
+        Type::Array(length, item) => {
+            visit_type(length, v);
+            visit_type(item, v);
+        }
+        Type::Slice(item) => {
+            visit_type(item, v);
+        }
+        Type::Integer(_, _) => {}
+        Type::Bool => {}
+        Type::String(length) => {
+            visit_type(length, v);
+        }
+        Type::FmtString(length, item) => {
+            visit_type(length, v);
+            visit_type(item, v);
+        }
+        Type::Unit => {}
+        Type::Tuple(items) => {
+            for typ in items {
+                visit_type(typ, v);
+            }
+        }
+        Type::DataType(typ, generics) => {
+            let typ = typ.borrow();
+            if let Some(fields) = typ.get_fields(generics) {
+                for (_, typ, _) in &fields {
+                    visit_type(typ, v);
+                }
+            } else if let Some(variants) = typ.get_variants(generics) {
+                let types = variants.iter().flat_map(|(_, types)| types);
+                for typ in types {
+                    visit_type(typ, v);
+                }
+            }
+        }
+        Type::Alias(typ, generics) => {
+            visit_type(&typ.borrow().get_type(generics), v);
+        }
+        Type::TypeVariable(type_variable)
+        | Type::NamedGeneric(NamedGeneric { type_var: type_variable, .. }) => {
+            match &*type_variable.borrow() {
+                TypeBinding::Bound(binding) => visit_type(binding, v),
+                TypeBinding::Unbound(_, _) => {}
+            }
+        }
+        Type::TraitAsType(_, _, trait_generics) => {
+            let types = trait_generics
+                .ordered
+                .iter()
+                .chain(trait_generics.named.iter().map(|named| &named.typ));
+            for typ in types {
+                visit_type(typ, v);
+            }
+        }
+        Type::CheckedCast { from, to } => {
+            visit_type(from, v);
+            visit_type(to, v);
+        }
```

Comment:
A checked cast is a cast from `from` to `to` so the resulting type itself is only `to`. If we visit both of these then it will throw off results, e.g. a count fields would return 2 with `from: Field, to: Field` instead of just 1.